lib/pure/asyncfutures.nim

```diff
@@ -32,27 +32,38 @@ type
 when not defined(release):
   var currentID = 0
 
-var callSoonProc {.threadvar.}: proc (cbproc: proc ()) {.gcsafe.}
+proc getCallSoonProc*(): (proc(cbproc: proc ()) {.gcsafe.}) {.deprecated.} = discard
+proc setCallSoonProc*(p: (proc(cbproc: proc ()) {.gcsafe.})) {.deprecated.} = discard
 
-proc getCallSoonProc*(): (proc(cbproc: proc ()) {.gcsafe.}) =
-  ## Get current implementation of ``callSoon``.
-  return callSoonProc
+var pendingCallbacks {.threadvar.}: seq[CallbackFunc]
 
-proc setCallSoonProc*(p: (proc(cbproc: proc ()) {.gcsafe.})) =
-  ## Change current implementation of ``callSoon``. This is normally called when dispatcher from ``asyncdispatcher`` is initialized.
-  callSoonProc = p
-
-proc callSoon*(cbproc: proc ()) =
+proc callSoon*(cbproc: CallbackFunc) =
   ## Call ``cbproc`` "soon".
   ##
-  ## If async dispatcher is running, ``cbproc`` will be executed during next dispatcher tick.
-  ##
-  ## If async dispatcher is not running, ``cbproc`` will be executed immediately.
-  if callSoonProc.isNil:
-    # Loop not initialized yet. Call the function directly to allow setup code to use futures.
-    cbproc()
-  else:
-    callSoonProc(cbproc)
+  ## The `cbproc` will be executed immediately unless `callSoon`
+  ## is called recursively. In such case `cbproc` will be called
+  ## as soon as currently handled `cbproc` is complete.
+  if pendingCallbacks.isNil:
+      pendingCallbacks = newSeqOfCap[CallbackFunc](8)
+  pendingCallbacks.add(cbproc)
+
+  # Every callback is added to pendingCallbacks, but in order to avoid recursion
+  # (some callback may result in a call to callSoon again) we start processing
+  # pendingCallbacks only upon the first addition, while still allowing for new
+  # callbacks to be appended to pendingCallbacks due to recursive reentrancy.
+
+  # More to think about. pendingCallbacks may grow as long as how many calls to
+  # callSoon are made during a single future completion cascade. That is unlikely
+  # to cause problems, but could be fixed by shrinking pendingCallbacks upon
+  # reaching some threshold.
+
+  if pendingCallbacks.len == 1:
+    var i = 0
+    while i < pendingCallbacks.len:
+      pendingCallbacks[i]()
```

Comment:
Could we end up in a situation where this line raises, causing `pendingCallbacks.len` to be something higher than "1" and thus the above branch fails forever, leading to all callbacks not completing?